Libs/GObject-2.0/Native/Classes/TypeDictionary.Registration.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace GObject.Native
+{
+    public static partial class TypeDictionary
+    {
+        // TODO: This stuff needs to be hidden behind a ITypeInitialiser interface
```

Comment:
I think we are starting to leave behind our prototype phase as we already know that we will get the code working, as it is working already in the develop branch. This is why I want to start with code as clean as possible now to avoid another rewrite not to far in the future.

Can you please cleanly separate the lazy loading and the TypeDictionary in different classes as both are different things which should not influence each other. 

One possible idea is that the TypeDictionary itself gets to be a static wrapper around some object which can resolve types. And then we can have different implementations of this object. One beeing lazy and another one expecting everything to be present at program startup. 
This would mean we actually have two different type dictionaries and the static wrapper decides which one to use.

Then we will probably see, that both implementations share some common code which we can refactor into the TypeDictionary itself to get the essential interface for "different kinds of loading types".

Reply:
The code is not really separate. E.g.: `RegisterSubclass` is part of `TypeRegistrar` but in case of "preregistered" subclasses there is no need for the `TypeRegistrar` to have this method.

If we have some kind of fallback mechanic it is not the task of the type dictionary to do this. The task of the type dictionary is to provide types. If a type is not found it is the task of the caller to decide what to do. Perhaps we need to add another layer on top of the type dictionary.

I try to adhere to the SOLID principles and I think the type dictionary is not following the single responsibility principle.

I'm still in the phase of internalize those principles but I think they help a lot creating better code.

If you prefer we can merge the code but it probably can not stay like it is now. On the other hand I think it is a good way to practice "clean code" if we work out the architecture for this together in this PR. What do you think?